_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,118 @@
+in memory are not known during compile time. This mechanism had a bug where, the Keccak-256 hash of
+the same memory content, but of different sizes were considered equal. The following is a simple
+example that illustrates the bug:
+
+```solidity
+contract C {
+  function bug() public returns (uint a, uint b) {
+    assembly {
+      mstore(0, 0)
+      // The optimizer computes the value at compile time:
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      a := keccak256(0, 32)
+      // The optimizer incorrectly uses the cached value
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      // instead of 
+      // 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
+      b := keccak256(0, 23)
+    }
+  }
+}
+```
+
+In the above example, both the Keccak-256 hashes can be computed at compile time. Here are two other
+cases where the hashes cannot be computed at compile time, yet the optimizer considered them equal
+and thus modified the semantics:
+
+```solidity
+contract C {
+  function bug(string memory s) public returns (bool ret) {
+    assembly {
+      let a := keccak256(s, 32)
+      let b := keccak256(s, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+```solidity
+contract C {
+  function bug() public view returns (bool ret) {
+    assembly {
+      let x := calldataload(0)
+      mstore(0, x)
+      mstore(0x20, x)
+      let a := keccak256(0, 4)
+      // even though the memory location is different,
+      // the 32-byte content is the same.
+      let b := keccak256(0x20, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+Specifically, `keccak256(mpos1, length1)` and `keccak256(mpos2, length2)` in some cases were
+considered equal if `length1` and `length2`, when rounded up to nearest multiple of 32 were the
+same, and when the memory contents at `mpos1` and `mpos2` can be deduced to be equal.
+
+You maybe **affected** if you compute multiple Keccak-256 hashes of the same content, but with
+different length inside inline assembly, and enabled the optimizer. 
+
+### Unaffected Cases
+
+You are unaffected in the following situations:
+
+- If your code uses `keccak256` with a length that is not known at compile time.
+- If your code only uses `keccak256` with length that is always a multiple of 32. In particular, the
+  Solidity compiler always uses multiples of 32 for storage slot computation.
+- If pairs of `keccak256` hashes are separated by certain instructions such as:
+
+  - An instruction that breaks the control flow, e.g., `jumpi`, `jump` etc. In high level code, this
+     would be function calls, if statements, loops, etc.
+  - An instruction that writes to memory other than `mstore`, e.g., `call`, `returndatacopy`, etc..
```

Comment:
Interesting! I thought `mstore` would invalidate the cache.

Reply:
`mstore` to a location that can be proven to be seperated by `32` does not invalidate `memory` or `keccak` cache:

```
			  let a := keccak256(0, 32)
			  mstore(64, 100)
			  let b := keccak256(0, 32)
			  ret := eq(a, b)
```

Reply:
Ah I see! Perhaps reword it as `invocations of mstore that certainly to a different memory location`? Because the current wording gives the impression that mstore to the same location also does not invalidate the cache. May be a reader should pay more attention here, but at least I didn't :-)